src/Autofac/Core/Registration/ServiceRegistrationInfo.cs

```diff
@@ -122,54 +144,88 @@ public bool IsRegistered
         }
 
         private bool Any =>
-            _defaultImplementations.Count > 0 ||
-            _sourceImplementations != null ||
-            _preserveDefaultImplementations != null;
+            Volatile.Read(ref _defaultImplementations).Count > 0 ||
+            Volatile.Read(ref _sourceImplementations) != null ||
+            Volatile.Read(ref _preserveDefaultImplementations) != null;
 
         public void AddImplementation(IComponentRegistration registration, bool preserveDefaults, bool originatedFromSource)
         {
             if (preserveDefaults)
             {
                 if (originatedFromSource)
                 {
-                    if (_sourceImplementations == null)
+                    // called inside of lock, just making List _sourceImplementations thread safe
+                    if (Volatile.Read(ref _sourceImplementations) == null)
                     {
-                        _sourceImplementations = new List<IComponentRegistration>();
+                        Volatile.Write(ref _sourceImplementations, new List<IComponentRegistration>());
                     }
 
-                    _sourceImplementations.Add(registration);
+                    var newSourceImplementations = new List<IComponentRegistration>(Volatile.Read(ref _sourceImplementations));
+                    newSourceImplementations.Add(registration);
+
+                    Volatile.Write(ref _sourceImplementations, newSourceImplementations);
                 }
                 else
                 {
-                    if (_preserveDefaultImplementations == null)
+                    // called inside of lock, just making List _preserveDefaultImplementations thread safe
+                    if (Volatile.Read(ref _preserveDefaultImplementations) == null)
                     {
-                        _preserveDefaultImplementations = new List<IComponentRegistration>();
+                        Volatile.Write(ref _preserveDefaultImplementations, new List<IComponentRegistration>());
                     }
 
-                    _preserveDefaultImplementations.Add(registration);
+                    var newPreserveDefaultImplementations = new List<IComponentRegistration>(Volatile.Read(ref _preserveDefaultImplementations));
```

Comment:
PR #919 attempts to optimize the amount of memory traffic going on in an effort to speed up things like mobile applications, where allocations cost time and battery. Having to allocate a whole new list every time here seems like it might counteract some of those optimizations. We might have to balance the lock contention improvement with the memory usage/optimization improvements.